### ISSUE TITLE
Asynchronous programming. Await the Future

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4624]
Asynchronous programming. Await the Future<br>
<br>
https://ift.tt/2IuGtHl<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<ins data-ad-slot="6474574513" data-full-width-responsive="true" data-ad-client="ca-pub-4665632381152577" data-ad-format="auto"></ins>
<p>This is the third post of a series about asynchronous programming. The whole series tries to answer the simple question: "What is asynchrony?".</p>
<p>At first, when I just started digging into the question - I thought that I know what it is. It turned out that I didn't know a clue about what asynchrony is all about. So, let's find out!</p>
<p>Whole series:</p>
<ul>
<li><a href="https://luminousmen.com/post/asynchronous-programming-blocking-and-non-blocking">Asynchronous programming. Blocking I/O and non-blocking I/O</a></li>
<li><a href="https://luminousmen.com/post/asynchronous-programming-cooperative-multitasking">Asynchronous programming. Cooperative multitasking</a></li>
<li><a href="https://luminousmen.com/post/asynchronous-programming-await-the-future">Asynchronous programming. Await the Future</a></li>
<li><a href="https://luminousmen.com/post/asynchronous-programming-python3.5">Asynchronous programming. Python3.5+</a></li>
</ul>
<p>Some applications implement parallelism using multiple processes instead of multiple threads(<a href="https://luminousmen.com/post/asynchronous-programming-blocking-and-non-blocking">first post</a>). Although the programming details are different, conceptually it is the same model and in this post, I'm talking in terms of <strong>threads</strong> but you can easily change it to processes.</p>
<p>Also here we will be talking only in terms of explicit cooperative multitasking &mdash; <strong>callbacks</strong>, because it's the most common and widely used option for asynchronous frameworks implementations.</p>
<hr>
<p>The <a href="https://insights.stackoverflow.com/survey/2018/">most common activity of modern applications is to deal with input and output</a>, rather than a lot of number-crunching. The problem with using input/output(I/O) functions is that they are <strong><a href="https://luminousmen.com/post/asynchronous-programming-blocking-and-non-blocking">blocking</a></strong>. The actual write to a hard disk or reading from network takes an extremely long time compared to the <a href="https://www.slideshare.net/ikewu83/dean-keynoteladis2009-4885081/24">speed of the CPU</a>. The functions don&rsquo;t finish until the task is done so meanwhile, your application is doing nothing. For applications which require high performance, this is a major roadblock as other activities and other I/O operations are kept waiting.</p>
<p>One of the standard solutions is to use threads. Each blocking I/O operation is started in a separate thread. When the blocking function gets invoked in the thread, the processor can schedule another thread to run, which actually needs the CPU.</p>
<p>In this post, we will be talking about the concepts of synchrony and asynchrony in general.</p>
<h2>Synchronicity</h2>
<p>In this concept, a <strong>thread is assigned to a single task</strong> and starts working on it. When a task is completed, the thread takes the next task and does the same: it performs all its commands one after the other to complete one specified task. In this system, a thread cannot leave the task halfway and move on to the next. Because of this, we can know for sure: whenever and wherever a function is executed &mdash; it <strong>cannot be set on hold</strong> and will be fully completed before starting to execute another one (which can change the data with which the current function works).</p>
<h3>Single-threaded</h3>
<p>If the system is executed single-threadedly and several tasks are connected with it, then they will be executed in this one thread sequentially one after the other.</p>
<p><img alt="Single-threaded" src="https://luminousmen.com/media/asynchronous-programming-await-the-future-1.jpg"></p>
<p>And if the tasks are always performed in a definite order, the implementation of a later task can assume that all earlier tasks have finished without errors, with all their output available for use &mdash; a definite simplification in logic.</p>
<p>If one of the commands is slow than the whole system will be waiting for this command to finish &mdash; there is no way around it.</p>
<h3>Multi-threaded</h3>
<p>In a multi-threaded system, the principle is preserved &mdash; one thread is assigned to one task and works on it until it is completed.</p>
<p>But in this system, each task is performed in a separate thread of control. The <strong>threads are managed by the operating system</strong> and may, on a system with multiple processors or multiple cores, run in parallel, or may run be interleaved together on a single processor.</p>
<p>Only now we have more than one thread and the tasks (not one task, but several different tasks) can be executed in parallel. Usually, tasks differ in the duration of processing and in fact, the thread that has finished working on one of its tasks can go to the next one.</p>
<p><img alt="Multithreaded" src="https://luminousmen.com/media/asynchronous-programming-await-the-future-3.jpg"></p>
<p>Multithreaded programs are more complicated, and typically more error-prone, they include common troublesome issues: race-conditions, dead-locks and resource starvation.</p>
<h2>Asynchrony</h2>
<p>The other approach uses another style, which is the asynchronous, <strong>non-blocking style</strong>. Asynchronous is a style of concurrent programming, but <a href="https://luminousmen.com/post/concurrency-and-parallelism-are-different">it is not parallelism</a>.</p>
<p>Most modern operating systems provide <a href="https://linux.die.net/man/2/eventfd2">event notification subsystems</a>. For example, a normal <code>read</code> call on a socket would block until the sender actually sent something. Instead, the application can request the operating system to watch the socket and put an event notification in the queue. The application can inspect the events at its convenience(perhaps doing some number crunching before to use the processor to the maximum) and grab the data. It is asynchronous because <strong>the application expressed interest at one point, then used the data at another point</strong> (in time and space). It is non-blocking because the application thread was free to do other tasks.</p>
<p><img alt="Asynchronous" src="https://luminousmen.com/media/asynchronous-programming-await-the-future-2.jpg"></p>
<p>Asynchronous code removes the <strong>blocking</strong> operation from the main application thread, so that it continues to be executed, but sometime later(or maybe somewhere else), and the handler can go further. Simply put, the main thread sets the task and transfers it's execution to some time later(or to another independent thread).</p>
<h3>Asynchrony and context switching</h3>
<p>While asynchronous programming can prevent all these issues, it was actually designed for an entirely different problem: CPU context switching. When you have multiple threads running, each CPU core can still only run one thread at a time. In order to allow all threads/processes to share resources, <strong>the CPU switch context very often</strong>. To oversimplify things, the CPU, at a random interval, saves all the context info of a thread and switches to another thread. The CPU is constantly switching between your threads in non-deterministic intervals. Threads are also resources, they are not free.</p>
<p><strong>Asynchronous programming is essentially cooperative multitasking with user-space threading, where the application manages the threads and context switching rather than the CPU</strong>. Basically, in an asynchronous world, context is switched only at defined switch points rather than in non-deterministic intervals.</p>
<h3>Comparison</h3>
<p>Compared to the synchronous model, the asynchronous model performs best when:</p>
<ul>
<li>There are a large number of tasks so there is likely always at least one task that can make progress;</li>
<li>The tasks perform lots of I/O, causing a synchronous program to waste lots of time blocking when other tasks could be running;</li>
<li>The tasks are largely independent of one another so there is little need for inter-task communication (and thus for one task to wait upon another).</li>
</ul>
<p>These conditions almost perfectly characterize a typical busy server (like a web server) in a client-server environment. Each task represents one client request with I/O in the form of receiving the request and sending the reply. A server implementation is a prime candidate for the asynchronous model, which is why <a href="https://github.com/twisted/twisted">Twisted</a> and <a href="https://nodejs.org/">Node.js</a>, among other asynchronous server libraries, have grown so much popularity in recent years.</p>
<p><em>Why not just use more threads?</em> If one thread is blocking on an I/O operation, another thread can make progress, right? However, as the number of threads increases, your server may start to experience performance problems. With each new thread, there is some memory overhead associated with the creation and maintenance of the thread state. Another performance gain from the asynchronous model is that it avoids context switching &mdash; every time the OS transfers control over from one thread to another it has to save all the relevant registers, memory map, stack pointers, CPU context etc. so that the other thread can resume execution where it left off. The overhead of doing this can be quite significant.</p>
<h3>Event loop</h3>
<p><em>How does an event of new task arrival can reach the application if the execution thread is busy processing another task?</em> The fact is that the operating system has many threads and the code that actually interacts with the user is executed separately from our application and only sends messages to it.</p>
<p>And how are doing all the event-thread managing? <strong>In the event loop</strong>.</p>
<p><img alt="Event-loop" src="https://luminousmen.com/media/asynchronous-programming-await-the-future-4.jpg"></p>
<p>The event loop is exactly what it sounds like, there is a queue of events(where all the happened events are stored &mdash; it's called task queue on the picture above) and a loop that just constantly pulls these events off the queue and executes callbacks on these events(all execution goes on the call stack). API represents API for asynchronous functions calls like waiting for the response from client or database.</p>
<p>So all the operations go first into <strong>call stack</strong> than asynchronous commands goes into <strong>API</strong> and after they are done required callback goes into the <strong>task queue</strong> and then for the execution on call stack again.</p>
<p>Coordination of this process takes place in the event loop.</p>
<p>You see how is this different from the reactor pattern we talked <a href="https://luminousmen.com/post/asynchronous-programming-cooperative-multitasking">in last post</a>? Right - nothing.</p>
<p>When the event loop forms the central control flow construct of a program, as it often does, it may be termed the main loop or <strong>main event loop.</strong> This title is appropriate because such an event loop is at the highest level of control within the application.</p>
<p>In <strong>event-driven programming</strong>, an application expresses interest in certain events and respond to them when they occur. The responsibility of gathering events from the operating system or monitoring other sources of events is handled by the event loop, and the user can register callbacks to be invoked when an event occurs. The event-loop usually keeps running forever.</p>
<p>JS event loop concept explained: <a href="https://youtu.be/8aGhZQkoFbQ">What the heck is the event loop anyway? | Philip Roberts | JSConf EU</a></p>
<h2>Conclusion</h2>
<p>Summarizing the whole theoretical series:</p>
<ol>
<li>Asynchronous operations in applications can make it more efficient, and most importantly fast for the user.</li>
<li>Resource savings. OS threads are cheaper than processes but they are still very expensive to use one thread per task. It will be more efficient to reuse it &mdash; and that is what asynchronous programming is providing us.</li>
<li>This is one of the most important techniques for optimizing and scaling I/O-bound applications(yes &mdash; they will not help in case CPU-bound tasks)</li>
<li>Asynchronous programs are difficult for programmer to write and debug</li>
</ol>
<a href="https://www.buymeacoffee.com/OheL3Mfnp" target="blank">Support author </a><br>
<ins data-ad-slot="1561747547" data-full-width-responsive="true" data-ad-client="ca-pub-4665632381152577" data-ad-format="auto"></ins>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Blog | iamluminousmen https://luminousmen.com<br>
<br>
June 28, 2019 at 11:50AM
